File: mfg-lab/python/src/mfglab/lab/failure_map.py
"""failure_map.py — where does your solver stop working? Python twin of
``mfg-lab/lab/failure-map.js``.

You know your method works where you built it. You do not know where it STOPS.
This sweeps a parameter box and reports, per point, whether the solver reached
tolerance — plus, where adjacent samples disagree, the BRACKET containing the
change of behaviour.

It is called a bracket and never a boundary. A grid cannot see a failure region
thinner than its own spacing, and that limitation is an ASSUMPTION attached to
every certificate this module issues rather than a caveat in prose.

The outcome taxonomy is observed, not thresholded:

``ok``
    reached the tolerance it was given.
``stalled``
    still above tolerance but below 1 — converging, out of budget. "Give it
    more iterations."
``diverged``
    non-finite values, or a relative residual at or above 1, i.e. no better
    than the initial guess. "This will never work here."
``threw``
    the kernel raised.

The only number here is 1, and it is not a tolerance: it is the exact value of
the relative residual at the initial guess (see :mod:`.contract`), so a run
ending at or above it has gone backwards. Slow and divergent were ONE outcome
in the first version, and separating them was forced by measurement — the
Helmholtz fixture ended at residual 6.95e+24 and was reported as "stalled",
which tells a researcher to raise their iteration cap when no cap will help.
"""
from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Any, Mapping, Sequence

from ..certificate import Certificate, proved, refused
from .contract import ValidatedKernel, error_of, validate

OK, STALLED, DIVERGED, THREW = "ok", "stalled", "diverged", "threw"

FALSIFIERS = [
    "a parameter point inside the box where the solver does not reach tolerance",
    "a finer sample that finds failures this spacing stepped over — the reason this reports brackets and never boundaries",
    "a kernel that reports a residual it did not achieve (the map trusts the kernel on this one number)",
    "an iteration cap raised until a stalled point converges, which turns a \"failure\" into a cost",
]


def _assumptions_for(spacing_text: str) -> list[str]:
    return [
        f"a failure region thinner than the sample spacing ({spacing_text}) is INVISIBLE to this map — it samples, it does not prove",
        "the residual reported by the kernel is the residual it actually reached",
        '"stalled" is a property of the configuration swept, including its iteration cap, not of the scheme alone',
    ]


@dataclass
class Point:
    i: int
    j: int
    params: dict
    outcome: str
    detail: str = ""
    iters: Any = None
    residual: Any = None
    error: float | None = None


@dataclass
class Axis:
    name: str
    values: list[float]


@dataclass
class MapResult:
    code: str
    certificate: Certificate
    axes: list[Axis]
    points: list[Point]
    brackets: list[dict]
    counts: dict
    n: int
    tol: float
    samples: int = 0
    _by_index: dict = field(default_factory=dict, repr=False)


def _linspace(lo: float, hi: float, k: int) -> list[float]:
    if k == 1:
        return [(lo + hi) / 2]
    return [lo + (hi - lo) * i / (k - 1) for i in range(k)]


def _all_finite(v: Sequence[float]) -> bool:
    return all(isinstance(x, (int, float)) and math.isfinite(x) for x in v)


def _probe(K: ValidatedKernel, params: dict, n: int, tol: float) -> dict:
    """One parameter point. Every failure mode is classified rather than raised
    — a map that dies on its first bad point is a map of nothing."""
    try:
        out = K.solve(n=n, tol=tol, params=params)
    except Exception as e:  # noqa: BLE001 — classifying failures is the job
        return dict(outcome=THREW, detail=str(e), iters=None, residual=None, error=None)

    # NOT `out.get("u") or []`: numpy arrays raise on bool(), and a researcher's
    # Python kernel commonly returns np.ndarray for u. Truth-test None explicitly.
    raw = out.get("u")
    u = list(raw) if raw is not None else []
    if not u or not _all_finite(u):
        return dict(outcome=DIVERGED, detail="the returned solution contains non-finite values",
                    iters=out.get("iters"), residual=out.get("residual"), error=None)

    try:
        error = error_of(u, list(K.exact(n=n, params=params)), K.h(n), K.norm)
    except Exception:  # noqa: BLE001
        error = None
    if error is not None and not math.isfinite(error):
        return dict(outcome=DIVERGED, detail="the error against the reference solution is not finite",
                    iters=out.get("iters"), residual=out.get("residual"), error=None)

    r = out.get("residual")
    if r is None:
        return dict(outcome=OK, detail="", iters=out.get("iters"), residual=None, error=error)

    if not math.isfinite(r) or r >= 1:
        shown = f"{r:.3e}" if math.isfinite(r) else str(r)
        return dict(outcome=DIVERGED, iters=out.get("iters"), residual=r, error=error,
                    detail=f"the relative residual ended at {shown} — at or above its starting value, so the "
                           "iteration moved away from the solution rather than converging slowly")
    if r <= tol:
        return dict(outcome=OK, detail="", iters=out.get("iters"), residual=r, error=error)
    return dict(outcome=STALLED, iters=out.get("iters"), residual=r, error=error,
                detail=f"stopped at relative residual {r:.3e}, above the requested {tol:.3e} but still below 1 "
                       "— converging, too slowly for the budget it was given")


def map_box(
    kernel: Any,
    sweep: Mapping[str, Sequence[float]],
    samples: int = 12,
    n: int | None = None,
    tol: float = 1e-6,
    params: Mapping[str, Any] | None = None,
) -> MapResult:
    """Sweep one or two parameters over a box and classify every sample.

    Named ``map_box`` rather than ``map`` so it does not shadow the builtin —
    the JavaScript sibling is ``map`` because JS has no such conflict.
    """
    K = validate(kernel)
    if not sweep or not isinstance(sweep, Mapping):
        raise ValueError("map: sweep is required, e.g. {'k': [0.5, 6]}")
    names = list(sweep.keys())
    if not 1 <= len(names) <= 2:
        raise ValueError(f"map: sweep one or two parameters, got {len(names)}. "
                         "A three-dimensional box needs a different instrument.")
    for nm in names:
        r = sweep[nm]
        if (not isinstance(r, Sequence) or isinstance(r, str) or len(r) != 2
                or not all(isinstance(x, (int, float)) and math.isfinite(x) for x in r) or not r[0] < r[1]):
            raise ValueError(f"map: sweep.{nm} must be [lo, hi] with lo < hi")
    if isinstance(samples, bool) or not isinstance(samples, int) or samples < 2:
        raise ValueError("map: samples must be an integer >= 2; one sample per axis is a spot-check, "
                         "which is the thing this replaces")
    if n is None:
        n = K.levels[len(K.levels) // 2]

    axes = [Axis(nm, _linspace(sweep[nm][0], sweep[nm][1], samples)) for nm in names]
    spacing_text = ", ".join(
        f"{a.name} {(a.values[-1] - a.values[0]) / (samples - 1):.3g}" for a in axes)

    base = dict(K.params)
    base.update(params or {})
    points: list[Point] = []
    counts = {OK: 0, STALLED: 0, DIVERGED: 0, THREW: 0}

    outer = axes[0].values
    inner = axes[1].values if len(axes) > 1 else [None]
    for i, a in enumerate(outer):
        for j, b in enumerate(inner):
            p = dict(base)
            p[axes[0].name] = a
            if len(axes) > 1:
                p[axes[1].name] = b
            r = _probe(K, p, n, tol)
            counts[r["outcome"]] += 1
            points.append(Point(i=i, j=j, params=p, **r))

    def at(i: int, j: int) -> Point:
        return points[i * len(inner) + j]

    brackets: list[dict] = []
    for i in range(len(outer)):
        for j in range(len(inner)):
            a_pt = at(i, j)
            if i + 1 < len(outer):
                b_pt = at(i + 1, j)
                if b_pt.outcome != a_pt.outcome:
                    brackets.append(dict(axis=axes[0].name, from_=outer[i], to=outer[i + 1],
                                         fromOutcome=a_pt.outcome, toOutcome=b_pt.outcome,
                                         at={axes[1].name: inner[j]} if len(axes) > 1 else {}))
            if len(axes) > 1 and j + 1 < len(inner):
                b_pt = at(i, j + 1)
                if b_pt.outcome != a_pt.outcome:
                    brackets.append(dict(axis=axes[1].name, from_=inner[j], to=inner[j + 1],
                                         fromOutcome=a_pt.outcome, toOutcome=b_pt.outcome,
                                         at={axes[0].name: outer[i]}))

    total = len(points)
    bad = total - counts[OK]
    evidence = {
        "box": " x ".join(f"{nm} in [{sweep[nm][0]}, {sweep[nm][1]}]" for nm in names),
        "sampled points": total,
        "reached tolerance": counts[OK],
        "stalled / diverged / threw": f"{counts[STALLED]} / {counts[DIVERGED]} / {counts[THREW]}",
        "resolution n": n,
        "tolerance": tol,
    }
    common = dict(falsifier=FALSIFIERS, assumes=_assumptions_for(spacing_text),
                  provenance={"kernel": K.name, "samplesPerAxis": samples, "n": n, "tol": tol, "norm": K.norm})

    if bad == 0:
        code = "PROVED"
        worst = max(points, key=lambda p: p.iters or 0)
        cert = proved(claim=f"{K.name} reached its tolerance at every SAMPLED point of the box "
                            "(a sample, not a proof of the box)",
                      evidence={**evidence, "worst iteration count": worst.iters}, **common)
    else:
        code = "POINTS_FAILED"
        first_bad = next(p for p in points if p.outcome != OK)
        where = ", ".join(f"{nm}={first_bad.params[nm]:.4g}" for nm in names)
        if brackets:
            bracket_text = "The change of behaviour is bracketed by " + "; ".join(
                f"{b['axis']} in [{b['from_']:.4g}, {b['to']:.4g}] ({b['fromOutcome']} -> {b['toOutcome']})"
                for b in brackets[:3])
        else:
            bracket_text = ("No adjacent pair of samples changed outcome, so the whole sampled box behaves the "
                            "same way and no bracket exists.")
        cert = refused(claim=f"{K.name} reaches its tolerance across the box",
                       why=f"{bad} of {total} sampled points did not reach tolerance ({counts[STALLED]} stalled, "
                           f"{counts[DIVERGED]} diverged, {counts[THREW]} threw). First at {where}: "
                           f"{first_bad.detail or first_bad.outcome}. {bracket_text}",
                       evidence=evidence, **common)

    return MapResult(code=code, certificate=cert, axes=axes, points=points, brackets=brackets,
                     counts=counts, n=n, tol=tol, samples=samples)

File: mfg-lab/lab/failure-map.js
/* failure-map.js — where does YOUR solver stop working?

   THE PAIN. You know your method works, because you tested it where you built
   it. You do not know where it STOPS — and the parameter you are about to
   change for the next paper may be on the other side of a boundary nobody has
   drawn. The usual substitute is spot-checks: a handful of runs at values that
   felt interesting, none of which was chosen by the failure.

   This repository has already been caught by exactly that. A prose range in
   the artifact claimed a scheme "reaches 1e-9 in 50-160 iterations across the
   whole slider box". A 16-corner sweep measured 32-78 iterations AND found
   that FOUR OF SIXTEEN corners do not converge at all. The range was wrong in
   both directions, and the non-convergence was invisible until something swept
   the box instead of sampling the middle of it.

   WHAT THIS RETURNS, AND WHAT IT REFUSES TO CALL IT
   A sampled grid of outcomes over a parameter box, and — where the outcome
   changes between adjacent samples — the BRACKET containing the transition.
   It is called a bracket and never a boundary, because a grid sample cannot
   see a failure region thinner than its own spacing. That limitation is
   declared on every certificate this file produces; it is not a caveat in
   prose, it is an assumption attached to the claim.

   THE OUTCOME TAXONOMY IS OBSERVED, NOT THRESHOLDED
     ok        reached the tolerance it was given
     stalled   still above tolerance but below 1 — converging, and out of
               budget. "Give it more iterations."
     diverged  non-finite values, or a relative residual at or above 1 —
               no better than the initial guess. "This will never work here."
     threw     the kernel raised

   Nothing here compares against an invented constant. `ok` is the kernel's own
   residual against the tolerance the caller passed; `diverged` uses 1, which
   is not a tolerance but the exact value of the relative residual at the
   initial guess (see the contract) — a run ending there has gone backwards.
   The other two are structural.

   Slow and divergent were ONE outcome in the first version of this file, and
   separating them was forced by measurement rather than foresight: the
   Helmholtz fixture ended at residual 6.95e+24 and got reported as "stalled",
   which tells a researcher to raise their iteration cap when the truth is that
   no cap will help.

   ONE HONEST CONSEQUENCE, STATED UP FRONT. `stalled` is a property of YOUR
   configuration, not of the mathematics: a run that would converge in ten
   million iterations is stalled at a cap of ten thousand, and this map will
   say so. That is deliberate and matches the order study — the instrument
   characterises the setup in front of it, and says which knob it was holding.

   MIT licensed. Part of the MFG Lab. */
'use strict';

const { validate, errorOf } = require('./contract.js');
const { proved, refused } = require('../../eqcert/src/certificate.js');

const OK = 'ok', STALLED = 'stalled', DIVERGED = 'diverged', THREW = 'threw';

const FALSIFIERS = [
  'a parameter point inside the box where the solver does not reach tolerance',
  'a finer sample that finds failures this spacing stepped over — the reason this reports brackets and never boundaries',
  'a kernel that reports a residual it did not achieve (the map trusts the kernel on this one number)',
  'an iteration cap raised until a stalled point converges, which turns a "failure" into a cost'
];

function assumptionsFor(spacingText) {
  return [
    'a failure region thinner than the sample spacing (' + spacingText + ') is INVISIBLE to this map — it samples, it does not prove',
    'the residual reported by the kernel is the residual it actually reached',
    '"stalled" is a property of the configuration swept, including its iteration cap, not of the scheme alone'
  ];
}

function linspace(lo, hi, k) {
  if (k === 1) return [(lo + hi) / 2];
  const out = new Array(k);
  for (let i = 0; i < k; i++) out[i] = lo + (hi - lo) * i / (k - 1);
  return out;
}

function allFinite(v) {
  for (let i = 0; i < v.length; i++) if (!Number.isFinite(v[i])) return false;
  return true;
}

/* One parameter point. Every failure mode is caught and classified rather than
   thrown, because a map that dies on its first bad point is a map of nothing. */
function probe(K, params, n, tol) {
  let out;
  try {
    out = K.solve({ n, tol, params });
  } catch (e) {
    return { outcome: THREW, detail: e.message, iters: null, residual: null, error: null };
  }
  if (!out || !out.u || !allFinite(out.u))
    return { outcome: DIVERGED, detail: 'the returned solution contains non-finite values', iters: out && out.iters, residual: out && out.residual, error: null };

  let error = null;
  try {
    error = errorOf(out.u, K.exact({ n, params }), K.h(n), K.norm);
  } catch (e) { error = null; }
  if (error !== null && !Number.isFinite(error))
    return { outcome: DIVERGED, detail: 'the error against the reference solution is not finite', iters: out.iters, residual: out.residual, error: null };

  /* residual === null means a direct solver: there is no iteration to stall. */
  const r = out.residual;
  if (r === null || r === undefined) return { outcome: OK, detail: '', iters: out.iters, residual: null, error };

  /* DIVERGED vs STALLED, decided structurally rather than by a threshold.
     The contract asks for a RELATIVE residual, so 1 is exactly "no better than
     the initial guess". A run that ends at or above 1 has not converged slowly
     — it has moved away from the answer, or overflowed on the way. Slow and
     divergent are different answers to the question being asked, and lumping
     them together was the first version of this file: the Helmholtz fixture
     ended at residual 6.95e+24 and was reported as "stalled", which reads as
     "needs more iterations" when the truth is "this iteration will never work
     here". Measured, then fixed. */
  if (!Number.isFinite(r) || r >= 1)
    return { outcome: DIVERGED, detail: 'the relative residual ended at ' + (Number.isFinite(r) ? r.toExponential(3) : String(r)) +
             ' — at or above its starting value, so the iteration moved away from the solution rather than converging slowly',
             iters: out.iters, residual: r, error };

  if (r <= tol) return { outcome: OK, detail: '', iters: out.iters, residual: r, error };
  return {
    outcome: STALLED,
    detail: 'stopped at relative residual ' + r.toExponential(3) + ', above the requested ' + tol.toExponential(3) +
            ' but still below 1 — converging, too slowly for the budget it was given',
    iters: out.iters, residual: r, error
  };
}

/* opts:
     sweep    REQUIRED. {paramName: [lo, hi]} for one or two parameters.
     samples  per axis, default 12
     n        resolution to hold fixed, default the kernel's median level
     tol      iteration tolerance, default 1e-6
     params   base parameters; the swept ones are overwritten per point     */
function map(kernel, opts) {
  opts = opts || {};
  const K = validate(kernel);
  const sweep = opts.sweep;
  if (!sweep || typeof sweep !== 'object') throw new Error('map: sweep is required, e.g. {k: [0.5, 6]}');

  const names = Object.keys(sweep);
  if (names.length < 1 || names.length > 2)
    throw new Error('map: sweep one or two parameters, got ' + names.length + '. A three-dimensional box needs a different instrument.');
  for (const nm of names) {
    const r = sweep[nm];
    if (!Array.isArray(r) || r.length !== 2 || !Number.isFinite(r[0]) || !Number.isFinite(r[1]) || !(r[0] < r[1]))
      throw new Error('map: sweep.' + nm + ' must be [lo, hi] with lo < hi');
  }

  const samples = opts.samples === undefined ? 12 : opts.samples;
  if (!Number.isInteger(samples) || samples < 2)
    throw new Error('map: samples must be an integer >= 2; one sample per axis is a spot-check, which is the thing this replaces');
  const tol = opts.tol === undefined ? 1e-6 : opts.tol;
  const n = opts.n === undefined ? K.levels[Math.floor(K.levels.length / 2)] : opts.n;

  const axes = names.map(nm => ({ name: nm, values: linspace(sweep[nm][0], sweep[nm][1], samples) }));
  const spacingText = axes.map(a => a.name + ' ' + ((a.values[a.values.length - 1] - a.values[0]) / (samples - 1)).toPrecision(3)).join(', ');

  /* ---- sweep ---- */
  const base = Object.assign({}, K.params, opts.params || {});
  const points = [];
  const counts = { ok: 0, stalled: 0, diverged: 0, threw: 0 };

  const outer = axes[0].values;
  const inner = axes[1] ? axes[1].values : [null];
  for (let i = 0; i < outer.length; i++) {
    for (let j = 0; j < inner.length; j++) {
      const params = Object.assign({}, base);
      params[axes[0].name] = outer[i];
      if (axes[1]) params[axes[1].name] = inner[j];
      const r = probe(K, params, n, tol);
      counts[r.outcome]++;
      points.push(Object.assign({ i, j, params }, r));
    }
  }

  /* ---- transition brackets: adjacent samples whose outcome differs ----
     This is the answer to the question actually being asked. "Your solver
     stops reaching tolerance between k=3.0 and k=3.5" is usable; a grid of
     coloured cells is a picture of it. */
  const at = (i, j) => points[i * inner.length + j];
  const brackets = [];
  for (let i = 0; i < outer.length; i++) {
    for (let j = 0; j < inner.length; j++) {
      const a = at(i, j);
      if (i + 1 < outer.length) {
        const b = at(i + 1, j);
        if (b.outcome !== a.outcome) brackets.push({ axis: axes[0].name, from: outer[i], to: outer[i + 1], fromOutcome: a.outcome, toOutcome: b.outcome, at: axes[1] ? { [axes[1].name]: inner[j] } : {} });
      }
      if (axes[1] && j + 1 < inner.length) {
        const b = at(i, j + 1);
        if (b.outcome !== a.outcome) brackets.push({ axis: axes[1].name, from: inner[j], to: inner[j + 1], fromOutcome: a.outcome, toOutcome: b.outcome, at: { [axes[0].name]: outer[i] } });
      }
    }
  }

  /* ---- the certificate ---- */
  const total = points.length;
  const bad = total - counts.ok;
  const evidence = {
    'box': names.map(nm => nm + ' in [' + sweep[nm][0] + ', ' + sweep[nm][1] + ']').join(' x '),
    'sampled points': total,
    'reached tolerance': counts.ok,
    'stalled / diverged / threw': counts.stalled + ' / ' + counts.diverged + ' / ' + counts.threw,
    'resolution n': n,
    'tolerance': tol
  };
  const commonBase = {
    falsifier: FALSIFIERS, assumes: assumptionsFor(spacingText),
    provenance: { kernel: K.name, samplesPerAxis: samples, n, tol, norm: K.norm }
  };

  let certificate, code;
  if (bad === 0) {
    code = 'PROVED';
    const worst = points.reduce((w, p) => (p.iters || 0) > (w.iters || 0) ? p : w, points[0]);
    certificate = proved(Object.assign({}, commonBase, {
      claim: K.name + ' reached its tolerance at every SAMPLED point of the box (a sample, not a proof of the box)',
      evidence: Object.assign({}, evidence, { 'worst iteration count': worst.iters })
    }));
  } else {
    code = 'POINTS_FAILED';
    const firstBad = points.find(p => p.outcome !== OK);
    const where = names.map(nm => nm + '=' + (+firstBad.params[nm]).toPrecision(4)).join(', ');
    certificate = refused(Object.assign({}, commonBase, {
      claim: K.name + ' reaches its tolerance across the box',
      why: bad + ' of ' + total + ' sampled points did not reach tolerance (' + counts.stalled + ' stalled, ' +
           counts.diverged + ' diverged, ' + counts.threw + ' threw). First at ' + where + ': ' +
           (firstBad.detail || firstBad.outcome) + '. ' +
           (brackets.length ? 'The change of behaviour is bracketed by ' +
              brackets.slice(0, 3).map(b => b.axis + ' in [' + (+b.from).toPrecision(4) + ', ' + (+b.to).toPrecision(4) + '] (' + b.fromOutcome + ' -> ' + b.toOutcome + ')').join('; ')
            : 'No adjacent pair of samples changed outcome, so the whole sampled box behaves the same way and no bracket exists.'),
      evidence
    }));
  }

  return { code, certificate, axes, points, brackets, counts, n, tol, samples };
}

module.exports = { map, OK, STALLED, DIVERGED, THREW, FALSIFIERS };
